File: src/main/java/com/windea/java/utils/JsonUtils.java
package com.windea.java.utils;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonWriter;

import java.io.FileReader;
import java.io.FileWriter;
import java.util.Map;

public class JsonUtils {
	private JsonUtils() {
	}


	public static Map fromFile(String path) throws Exception {
		return fromFile(path, Map.class);
	}

	public static <T> T fromFile(String path, Class<T> type) throws Exception {
		var reader = new FileReader(path);
		T result = json().fromJson(reader, type);
		return result;
	}

	public static Map fromString(String string) {
		return fromString(string, Map.class);
	}

	public static <T> T fromString(String string, Class<T> type) {
		var result = json().fromJson(string, type);
		return result;
	}

	public static <T> void toFile(T data, String path) throws Exception {
		toFile(data, path, 2);
	}

	public static <T> void toFile(T data, String path, int indent) throws Exception {
		var writer = new JsonWriter(new FileWriter(path));
		var type = new TypeToken<T>() {}.getType();
		writer.setIndent(" ".repeat(indent));
		json().toJson(data, type, writer);
	}

	public static <T> String toString(T json) {
		return json().toJson(json);
	}


	private static Gson json() {
		return new Gson();
	}
}

File: src/main/kotlin/com/windea/kotlin/utils/JsonUtils.kt
package com.windea.kotlin.utils

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonWriter

import java.io.FileReader
import java.io.FileWriter

object JsonUtils {
	@Throws(Exception::class)
	fun fromFile(path: String): Map<*, *> {
		return fromFile(path, Map::class.java)
	}

	@Throws(Exception::class)
	fun <T> fromFile(path: String, type: Class<T>): T {
		val reader = FileReader(path)
		return json().fromJson(reader, type)
	}

	fun fromString(string: String): Map<*, *> {
		return fromString(string, Map::class.java)
	}

	fun <T> fromString(string: String, type: Class<T>): T {
		return json().fromJson(string, type)
	}

	@Throws(Exception::class)
	fun <T> toFile(data: T, path: String) {
		toFile(data, path, 2)
	}

	@Throws(Exception::class)
	fun <T> toFile(data: T, path: String, indent: Int) {
		val writer = JsonWriter(FileWriter(path))
		val type = object : TypeToken<T>() {}.type
		writer.setIndent(" ".repeat(indent))
		json().toJson(data, type, writer)
	}

	fun <T> toString(json: T): String {
		return json().toJson(json)
	}


	private fun json(): Gson {
		return Gson()
	}
}
